src/node_credentials.cc

```diff
@@ -30,13 +32,38 @@ bool linux_at_secure = false;
 namespace credentials {
 
 // Look up environment variable unless running as setuid root.
-bool SafeGetenv(const char* key, std::string* text) {
+bool SafeGetenv(const char* key, std::string* text, Environment* env) {
 #if !defined(__CloudABI__) && !defined(_WIN32)
   if (per_process::linux_at_secure || getuid() != geteuid() ||
       getgid() != getegid())
     goto fail;
 #endif
 
+  if (env != nullptr && !env->is_main_thread() &&
+      !env->process_object().IsEmpty()) {
+    TryCatch ignore_errors(env->isolate());
+    Local<Value> env_obj;
+    Local<Value> val;
+    if (!env->process_object()->Get(
```

Comment:
Hmm, I am not sure about accessing process.env here...can we make a separate store in C++ that is not a v8::Object, and proxy the get of the worker process.env to fallback with that instead?

Reply:
So … an earlier local version of this thing, I abstracted process.env to use a generic key-value-store class as a backend; and then implemented a store for the actual process object, and another for a generic string storage class. It turned out to be a lot more complex than I’d wanted, though.

Reply:
I think we just need to add a storage for user-overriden values to env, and search that before looking up the actual environment variables with system calls?  (It should be okay to use `thread_local_env` to get the env for current thread in the case of `SafeGetEnv`)Or am I missing something?

Reply:
> I think we just need to add a storage for user-overriden values to env, and search that before looking up the actual environment variables with system calls?

That would give us different behavior, right? We want a copy of the environment in Workers, not a fallback to the actual Environment

> It should be okay to use `thread_local_env` to get the env for current thread in the case of `SafeGetEnv`

I’d prefer to get rid of `thread_local_env`, it assumes that there’s one Environment per thread, which doesn’t have to be true in any way, it’s just a natural use case.

Reply:
> That would give us different behavior, right?

Right, I forgot it's a replacement/snapshot of the main threads' process.env again..Is it possible to pass the processed `options.env` to the JS Worker contstructor, and convert it to a `map<string, string>` that gets passed to the C++ worker constructor and stored in the environment? Then in the `LOAD_SCRIPT` handler, we just need to make sure the new `process.env` is a proxy that's backed by the C++ map?

> I’d prefer to get rid of thread_local_env, it assumes that there’s one Environment per thread, which doesn’t have to be true in any way, it’s just a natural use case.

Makes sense. Maybe we can just take an optional Environment in`SafeGetEnv`?

Reply:
> > I’d prefer to get rid of thread_local_env, it assumes that there’s one Environment per thread, which doesn’t have to be true in any way, it’s just a natural use case.

> Makes sense. Maybe we can just take an optional Environment in`SafeGetEnv`?

That is being added as an optional parameter here :)